### PR TITLE
fix #1866: race accessing topic replicas map

### LINKED ISSUE #1866: Fix race accessing topic replicas map
```
==================
WARNING: DATA RACE
Read by goroutine 8:
  runtime.mapiterinit()
      /home/dgnorton/src/go/src/runtime/hashmap.go:543 +0x0
  github.com/influxdb/influxdb/messaging.(*topic).encode()
      /home/dgnorton/go/src/github.com/influxdb/influxdb/messaging/broker.go:893 +0x777
  github.com/influxdb/influxdb/messaging.(*brokerFSM).MustApply()
      /home/dgnorton/go/src/github.com/influxdb/influxdb/messaging/broker.go:591 +0x2c5
  github.com/influxdb/influxdb/raft.func·008()
      /home/dgnorton/go/src/github.com/influxdb/influxdb/raft/log.go:1212 +0x1760
  github.com/influxdb/influxdb/raft.(*Log).applier()
      /home/dgnorton/go/src/github.com/influxdb/influxdb/raft/log.go:1220 +0x226

Previous write by goroutine 60:
  runtime.mapassign1()
      /home/dgnorton/src/go/src/runtime/hashmap.go:383 +0x0
  github.com/influxdb/influxdb/messaging.(*Replica).WriteTo()
      /home/dgnorton/go/src/github.com/influxdb/influxdb/messaging/broker.go:1013 +0x96f
  github.com/influxdb/influxdb/messaging.(*Handler).stream()
      /home/dgnorton/go/src/github.com/influxdb/influxdb/messaging/handler.go:101 +0x250
  github.com/influxdb/influxdb/messaging.(*Handler).ServeHTTP()
      /home/dgnorton/go/src/github.com/influxdb/influxdb/messaging/handler.go:54 +0x24e
  main.(*Handler).ServeHTTP()
      /home/dgnorton/go/src/github.com/influxdb/influxdb/cmd/influxd/handler.go:32 +0x349
  net/http.serverHandler.ServeHTTP()
      /home/dgnorton/src/go/src/net/http/server.go:1703 +0x1f6
  net/http.(*conn).serve()
      /home/dgnorton/src/go/src/net/http/server.go:1204 +0x1087

Goroutine 8 (running) created at:
  github.com/influxdb/influxdb/raft.(*Log).Open()
      /home/dgnorton/go/src/github.com/influxdb/influxdb/raft/log.go:275 +0x705
  github.com/influxdb/influxdb/messaging.(*Broker).Open()
      /home/dgnorton/go/src/github.com/influxdb/influxdb/messaging/broker.go:102 +0x36f
  main.openBroker()
      /home/dgnorton/go/src/github.com/influxdb/influxdb/cmd/influxd/run.go:219 +0x10e
  main.Run()
      /home/dgnorton/go/src/github.com/influxdb/influxdb/cmd/influxd/run.go:45 +0x447
  main.execRun()
      /home/dgnorton/go/src/github.com/influxdb/influxdb/cmd/influxd/main.go:118 +0x8e1
  main.main()
      /home/dgnorton/go/src/github.com/influxdb/influxdb/cmd/influxd/main.go:70 +0x4a9

Goroutine 60 (running) created at:
  net/http.(*Server).Serve()
      /home/dgnorton/src/go/src/net/http/server.go:1751 +0x3cd
  net/http.Serve()
      /home/dgnorton/src/go/src/net/http/server.go:1606 +0xdc
  main.func·003()
      /home/dgnorton/go/src/github.com/influxdb/influxdb/cmd/influxd/run.go:56 +0xb2
==================
```

### PR DESCRIPTION
**Pull request by dgnorton**

### CODE REVIEW

**Comment by toddboom** [comment]
:+1:

File: messaging/broker.go [hunk 1]
```diff
@@ -762,9 +762,24 @@ type topic struct {
...
+func (t *topic) addReplica(r *Replica) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	t.replicas[r.id] = r
+}
+
+// replica returns the replica with the specified ID.
+func (t *topic) replica(id uint64) *Replica {
+	t.mu.RLock()
+	defer t.mu.Unlock()
```
Comment (otoolep) [inline messaging/broker.go]: Should be `RUnlock`.

**Comment by otoolep** [comment]
Does this actually clean up the output `-race`. Looks like it should.

File: messaging/broker.go (hunk 1 above)
Comment (dgnorton) [inline messaging/broker.go]: Good catch.

**Comment by dgnorton** [comment]
It cleans up one of the `-race` problems but there were several.

**Pull request merged**